src/main/java/mate/academy/rickandmorty/service/impl/CharacterServiceImpl.java

```
@@ -0,0 +1,69 @@
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional
+@RequiredArgsConstructor
+public class CharacterServiceImpl implements CharacterService {
+
+    private final CharacterRepository characterRepository;
+    private final RickAndMortyClient rickAndMortyClient;
+    private final Random random = new Random();
+
+    @Override
+    public void loadInitialData() {
+        System.out.println("loadInitialData() виконується...");
+        List<ExternalCharacterDto> externalCharacterDto = rickAndMortyClient.fetchAllCharacters();
+        System.out.println("Отримано персонажів: " + externalCharacterDto.size());
+
+        List<Character> toSaveDataEntity = externalCharacterDto.stream()
+                .map(dto -> {
+                    Character character = new Character();
+                    character.setExternalId(dto.id().toString());
+                    character.setName(dto.name());
+                    character.setStatus(dto.status());
+                    character.setGender(dto.gender());
+                    return character;
+                })
+                .toList();
+
+        characterRepository.saveAll(toSaveDataEntity);
+        System.out.println("Збережено у БД " + toSaveDataEntity.size() + " персонажів.");
+    }
+
+    @Override
+    public List<CharacterResponseDto> searchByName(String name) {
+        return (name == null || name.isBlank())
+                ? List.of() :
+                characterRepository.findByNameContainingIgnoreCase(name).stream()
+                        .map(this::toDto)
+                        .toList();
+    }
+
+    @Override
+    public CharacterResponseDto getRandomCharacter() {
+        List<Character> characters = characterRepository.findAll();
+        if (characters.isEmpty()) {
+            return null;
+        }
+        Character characterEntity = characters.get(random.nextInt(characters.size()));
```

The current random selection implementation loads all characters into memory (findAll) and then selects one. For large datasets this is inefficient. Use repository.count() to get total and then fetch a single row using pagination or a native query with LIMIT 1 OFFSET. Also handle the empty-database case explicitly (return Optional or throw a custom exception mapped to 404) rather than returning null. This addresses checklist item #3.15 and avoids loading entire table into memory.